packages/babel-parser/src/plugins/typescript/scope.ts

```diff
@@ -143,14 +143,16 @@ export default class TypeScriptScopeHandler extends ScopeHandler<TypeScriptScope
   }
 
   checkLocalExport(id: N.Identifier) {
-    const topLevelScope = this.scopeStack[0];
     const { name } = id;
-    if (
-      !topLevelScope.types.has(name) &&
-      !topLevelScope.exportOnlyBindings.has(name) &&
-      !this.hasImport(name)
-    ) {
-      super.checkLocalExport(id);
+
+    if (this.hasImport(name)) return;
+
+    const len = this.scopeStack.length;
+    for (let i = 0; i < len; i++) {
```

Comment:
```suggestion
    for (let i = len - 1; i >= 0; i--) {
```
Nit: The scope stack is meant to be enumerated from last to the first, though the behaviour is not changed.

Reply:
Maybe starting at 0 is a little better for performance? But I don't mind going from the back to the front.

Reply:
We could also use
```js
if (this.scopeStack.some(scope => scope.types.has(name) || scope.exportOnlyBindings.has(name))) return;
```

since `some` makes it more clear that the order doesn't matter but we are just checking if there exists a scope with that property.

Otherwise I prefer, as @JLHwung, to iterate them in the "expected" order, so that if in the future we add more logic to the loop we don't risk a bug caused by the wrong iteration order.